include/grpc/event_engine/event_engine.h

```diff
@@ -144,40 +154,67 @@ class EventEngine {
-  using OnConnectCallback = std::function<void(absl::Status, Endpoint*)>;
+  /// Called when a new connection is established.
+  ///
+  /// If the connection attempt was not successful, implementations should pass
+  /// the appropriate statuses to \a on_read. For example, callbacks might
+  /// expect to receive DEADLINE_EXCEEDED statuses when appropriate, or
+  /// CANCELLED statuses on EventEngine shutdown.
+  using OnConnectCallback =
+      std::function<void(absl::Status, std::unique_ptr<Endpoint>)>;
 
   /// An EventEngine Listener listens for incoming connection requests from gRPC
   /// clients and initiates request processing once connections are established.
   class Listener {
    public:
-    /// A callback handle, used to cancel a callback. Called when the listener
-    /// has accepted a new client connection. This callback takes ownership of
-    /// the Endpoint and is responsible its destruction.
-    using AcceptCallback = std::function<void(absl::Status, Endpoint*)>;
-
-    virtual ~Listener() = 0;
-
-    // TODO(hork): define return status codes
-    /// Bind an address/port to this Listener. It is expected that multiple
-    /// addresses/ports can be bound to this Listener before Listener::Start has
-    /// been called. Returns the bound port or an error status.
+    /// Called when the listener has accepted a new client connection.
+    using AcceptCallback = std::function<void(std::unique_ptr<Endpoint>)>;
+    virtual ~Listener() = default;
+    /// Bind an address/port to this Listener.
+    ///
+    /// It is expected that multiple addresses/ports can be bound to this
+    /// Listener before Listener::Start has been called. Returns either the
+    /// bound port or an appropriate error status.
     virtual absl::StatusOr<int> Bind(const ResolvedAddress& addr) = 0;
     virtual absl::Status Start() = 0;
   };
 
-  // TODO(hork): define status codes for the callback
-  // TODO(hork): define return status codes
-  // TODO(hork): document status arg meanings for on_accept and on_shutdown
-  /// Factory method to create a network listener.
+  /// Factory method to create a network listener / server.
+  ///
+  /// Once a \a Listener is created and started, the \a on_accept callback will
+  /// be called once asynchronously for each established connection. Note that
+  /// unlike other callbacks, there is no status code parameter since the
+  /// callback will only be called in healthy scenarios where connections can be
+  /// accepted.
+  ///
+  /// \a on_shutdown will be called exactly once, with either an OK status for a
+  /// normal \a Listener or \a EventEngine shutdown, or a non-OK status if the
+  /// Listener or gRPC encountered an error.
+  ///
+  /// This method may return a non-OK status immediately if an error was
+  /// encountered in any synchronous steps required to create a \a Listener. If
+  /// this happens, neither \a on_shutdown nor \a on_accept callbacks will have
```

Comment:
This seems a little confusing, in light of the statement in the previous paragraph that on_shutdown will be called exactly once.  I suggest rewriting these two paragraphs as follows:

"""
This method may return a non-OK status immediately if an error was encountered in any synchronous steps required to create the Listener.  In this case, \a on_shutdown will never be called.

If this method returns a Listener, then \a on_shutdown will be invoked exactly once, when the Listener is shut down.  The status passed to it will indicate if there was a problem during shutdown.
"""

I don't think it's necessary to say that on_accept won't be called if there is no Listener, because if there is no Listener, then there are by definition no connections being accepted.

Also, I didn't say that the Listener will be shut down when the EventEngine is shut down, because I thought we said that EventEngine shutdown would wait for all current activity to complete, and the Listener is one such activity.  In other words, I think EventEngine shutdown would not complete until the Listener is shut down.

Reply:
Thanks for the improvement and the correction. You're right, we had decided that the EventEngine's shutdown will wait for the Listener to complete its business - a soft shutdown operation.